### PR TITLE
Disable track changing buttons in the music controller depending on the current screen 

### PR DESCRIPTION
**Pull request by marco-stephan**
Video: https://streamable.com/j24e9 (Not up-to-date; the playlist overlay now stays hidden when entering, for example, the result screen)

- [X] Depends on https://github.com/ppy/osu-framework/pull/844

### CODE REVIEW

File: osu.Game/Screens/OsuScreen.cs [hunk 1]
```diff
@@ -29,6 +29,8 @@ public abstract class OsuScreen : Screen
 
         internal virtual bool AllowRulesetChange => true;
 
+        internal virtual bool CanBeatmapChange => true;
```
Comment (DrabWeb) [inline osu.Game/Screens/OsuScreen.cs]: Maybe `AllowBeatmapChange` to be more consistent with `AllowRulesetChange`.

File: osu.Game/OsuGameBase.cs [hunk 2]
```diff
@@ -45,6 +46,8 @@ public class OsuGameBase : Framework.Game, IOnlineComponent
 
         public readonly Bindable<WorkingBeatmap> Beatmap = new Bindable<WorkingBeatmap>();
 
+        public readonly Bindable<OsuScreen> CurrentScreen = new Bindable<OsuScreen>();
```
Comment (DrabWeb) [inline osu.Game/OsuGameBase.cs]: I really don't like exposing this, would suggest instead exposing something like `AllowBeatmapChange` on the `MusicController` and setting that in `screenChanged`.

**Review by DrabWeb — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/MusicController.cs [hunk 3]
```diff
@@ -363,6 +406,9 @@ protected override void PopIn()
 
             FadeIn(transition_length, EasingTypes.OutQuint);
             dragContainer.ScaleTo(1, transition_length, EasingTypes.OutElastic);
+
+            if(Alpha == 0)
```
Comment (smoogipooo) [inline osu.Game/Overlays/MusicController.cs]: Spacing.

File: osu.Game/Overlays/MusicController.cs [hunk 4]
```diff
@@ -53,6 +57,39 @@ public class MusicController : FocusedOverlayContainer
         private Container dragContainer;
         private Container playerContainer;
 
+        private bool showPlaylistOnceAvailable;
+
+        private bool allowBeatmapChange = true;
+
+        public bool AllowBeatmapChange
```
Comment (smoogipooo) [inline osu.Game/Overlays/MusicController.cs]: This should probably be a bindable, and use e.g. `prevButton.BindTo(AllowBeatmapChange)`.

File: osu.Game/OsuGame.cs [hunk 5]
```diff
@@ -327,6 +327,8 @@ private void screenChanged(Screen newScreen)
             }
 
             ScreenChanged?.Invoke(newScreen);
+
+            musicController.AllowBeatmapChange = currentScreen.AllowBeatmapChange;
```
Comment (smoogipooo) [inline osu.Game/OsuGame.cs]: And here, this could be `musicController.AllowBeatmapChange.BindTo(currentScreen.AllowBeatmapChange)`.

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game/Overlays/MusicController.cs [hunk 6]
```diff
@@ -199,17 +205,42 @@ private void load(OsuGameBase game, OsuColour colours, LocalisationEngine locali
...
         protected override void LoadComplete()
         {
             beatmapBacking.ValueChanged += beatmapChanged;
+            beatmapBacking.DisabledChanged += beatmapDisabledChanged;
             beatmapBacking.TriggerChange();
 
             base.LoadComplete();
         }
 
+        private void beatmapDisabledChanged(bool newBeatmapBackingDisabled)
```
Comment (peppy) [inline osu.Game/Overlays/MusicController.cs]: just call the bool `disabled`

File: osu.Game/Screens/Ranking/Results.cs [hunk 7]
```diff
@@ -280,5 +280,7 @@ private void load(OsuColour colours)
 
             modeChangeButtons.Current.TriggerChange();
         }
+
+        internal override bool AllowBeatmapChange => false;
```
Comment (peppy) [inline osu.Game/Screens/Ranking/Results.cs]: put this at the top of the file with other `internal override`s

File: osu.Game/Graphics/UserInterface/IconButton.cs [hunk 8]
```diff
@@ -80,16 +82,44 @@ private void load(OsuColour colours)
...
+            if (!Enabled.Value)
+                return true;
+
             hover.FadeIn(500, EasingTypes.OutQuint);
             return base.OnHover(state);
         }
 
         protected override void OnHoverLost(InputState state)
         {
+            if (!Enabled.Value)
```
Comment (peppy) [inline osu.Game/Graphics/UserInterface/IconButton.cs]: you probably don't want this.

File: osu.Game/Graphics/UserInterface/IconButton.cs [hunk 9]
```diff
@@ -102,12 +132,18 @@ protected override bool OnClick(InputState state)
...
+                return true;
+
             content.ScaleTo(0.75f, 2000, EasingTypes.OutQuint);
             return base.OnMouseDown(state, args);
         }
 
         protected override bool OnMouseUp(InputState state, MouseUpEventArgs args)
         {
+            if (!Enabled.Value)
```
Comment (peppy) [inline osu.Game/Graphics/UserInterface/IconButton.cs]: nor this

File: osu.Game/Overlays/MusicController.cs [hunk 10]
```diff
@@ -55,6 +59,8 @@ public class MusicController : OsuFocusedOverlayContainer
         private Container dragContainer;
         private Container playerContainer;
 
+        private bool showPlaylistOnceAvailable;
```
Comment (peppy) [inline osu.Game/Overlays/MusicController.cs]: i don't get this

**Review by peppy — APPROVED** [review]

**Pull request merged**